File: BAEKJOON/1963 - 소수 경로/소수 경로.cs
using System;
using System.Collections.Generic;

class Program
{
    // 1000 이상 9999 이하의 소수를 판별하기 위한 배열
    static bool[] is_prime = new bool[10000];

    // 에라토스테네스의 체를 사용하여 소수를 구하는 함수
    static void SieveOfEratosthenes() {
        for (int i = 0; i < 10000; i++) is_prime[i] = true; // 배열을 true로 초기화
        is_prime[0] = is_prime[1] = false; // 0과 1은 소수가 아님

        for (int i = 2; i <= Math.Sqrt(9999); i++) { // 2부터 9999의 제곱근까지 반복
            if (is_prime[i]) { // i가 소수인 경우
                for (int j = i * i; j < 10000; j += i) { // i의 배수를 모두 소수가 아님으로 설정
                    is_prime[j] = false;
                }
            }
        }
    }

    // 한 자리만 바꿔서 나올 수 있는 이웃 소수들을 반환하는 함수
    static List<int> GetNeighbors(int num) {
        List<int> neighbors = new List<int>();
        char[] numArray = num.ToString().ToCharArray(); // 숫자를 문자열로 변환

        for (int i = 0; i < 4; i++) { // 각 자리수에 대해
            char original = numArray[i]; // 원래 숫자를 저장

            for (char d = '0'; d <= '9'; d++) { // 각 자리수를 0부터 9로 바꾸어 봄
                if (numArray[i] != d) { // 다른 숫자로 변경 시도
                    numArray[i] = d; // 해당 자리수를 변경
                    int new_num = int.Parse(new string(numArray)); // 새로운 숫자를 정수로 변환

                    // 1000 이상이고 소수인 경우만 이웃으로 추가
                    if (new_num >= 1000 && is_prime[new_num]) {
                        neighbors.Add(new_num);
                    }
                }
            }

            numArray[i] = original; // 원래 자리수로 복원
        }

        return neighbors; // 가능한 이웃 소수들을 반환
    }

    // BFS를 사용하여 최소 변환 횟수를 계산하는 함수
    static int BFS(int start, int end) {
        if (start == end) return 0; // 시작과 목표가 같으면 0을 반환

        Queue<Tuple<int, int>> q = new Queue<Tuple<int, int>>(); // 큐에는 현재 소수와 그 소수까지의 변환 횟수를 저장
        bool[] visited = new bool[10000]; // 각 숫자에 대해 방문 여부를 저장

        q.Enqueue(new Tuple<int, int>(start, 0)); // 시작 소수와 0번째 변환 횟수를 큐에 추가
        visited[start] = true; // 시작 소수 방문 처리

        while (q.Count > 0) {
            var current = q.Dequeue(); // 현재 소수와 변환 횟수를 큐에서 꺼냄
            int currentNum = current.Item1;
            int steps = current.Item2;

            List<int> neighbors = GetNeighbors(currentNum); // 이웃 소수들을 구함

            foreach (int neighbor in neighbors) { // 이웃 소수들을 하나씩 확인
                if (!visited[neighbor]) { // 방문하지 않은 소수인 경우
                    if (neighbor == end) return steps + 1; // 목표 소수에 도달한 경우 변환 횟수 반환
                    visited[neighbor] = true; // 방문 처리
                    q.Enqueue(new Tuple<int, int>(neighbor, steps + 1)); // 이웃 소수를 큐에 추가
                }
            }
        }

        return -1; // 목표 소수에 도달할 수 없는 경우 -1 반환
    }

    static void Main() {
        int T = int.Parse(Console.ReadLine()); // 테스트 케이스 수 입력

        SieveOfEratosthenes(); // 소수를 미리 계산

        while (T-- > 0) {
            string[] input = Console.ReadLine().Split(); // 시작 소수와 목표 소수를 입력받음
            int start = int.Parse(input[0]);
            int end = int.Parse(input[1]);

            int result = BFS(start, end); // BFS를 통해 변환 횟수 계산

            if (result == -1)
                Console.WriteLine("Impossible"); // 변환이 불가능한 경우
            else
                Console.WriteLine(result); // 변환이 가능한 경우 최소 횟수 출력
        }
    }
}

File: BAEKJOON/1963 - 소수 경로/소수 경로.py
from collections import deque  # deque를 사용하여 BFS(너비 우선 탐색)를 구현하기 위해 collections 모듈에서 deque를 가져옴

def sieve_of_eratosthenes():  # 에라토스테네스의 체 알고리즘을 사용하여 1000부터 9999까지의 소수를 미리 구하는 함수
    is_prime = [True] * 10000  # 숫자 0부터 9999까지 소수 여부를 저장하는 배열, 처음엔 모두 소수라고 가정
    is_prime[0] = is_prime[1] = False  # 0과 1은 소수가 아니므로 False로 설정
    
    for i in range(2, int(10000**0.5) + 1):  # 2부터 10000의 제곱근까지 반복, 소수 여부를 판단
        if is_prime[i]:  # i가 소수인 경우
            for j in range(i * i, 10000, i):  # i의 배수들을 소수가 아니라고 표시
                is_prime[j] = False  # i의 배수들은 소수가 아니므로 False로 설정
                
    primes = [i for i in range(1000, 10000) if is_prime[i]]  # 1000 이상 9999 이하의 소수 리스트 생성
    return is_prime  # 소수 여부를 나타내는 배열 반환

def get_neighbors(num, is_prime):  # 주어진 숫자에서 한 자리만 바꿔서 나오는 네 자리 소수를 찾는 함수
    neighbors = []  # 이웃 소수를 저장할 리스트
    num_str = str(num)  # 숫자를 문자열로 변환하여 각 자리 수에 접근
    
    for i in range(4):  # 네 자리 수의 각 자리마다 반복
        for d in '0123456789':  # 각 자리 숫자를 0부터 9까지 대입
            if num_str[i] != d:  # 현재 자리 숫자와 다른 숫자로만 변경
                new_num = int(num_str[:i] + d + num_str[i+1:])  # 해당 자리를 새로운 숫자로 변경한 새로운 숫자 생성
                if new_num >= 1000 and is_prime[new_num]:  # 1000 이상이며 소수인 경우
                    neighbors.append(new_num)  # 이웃 소수 리스트에 추가
                    
    return neighbors  # 가능한 이웃 소수 리스트 반환

def bfs(start, end, is_prime):  # BFS를 사용해 최소 변환 횟수를 찾는 함수
    if start == end:  # 시작 소수와 끝 소수가 같으면 0 반환
        return 0
    
    queue = deque([(start, 0)])  # 큐에 시작 소수와 변환 횟수 0을 넣음
    visited = [False] * 10000  # 각 숫자가 방문되었는지 여부를 저장할 배열
    visited[start] = True  # 시작 소수는 방문했으므로 True로 설정
    
    while queue:  # 큐가 빌 때까지 반복
        current, steps = queue.popleft()  # 큐에서 현재 소수와 변환 횟수를 꺼냄
        
        for neighbor in get_neighbors(current, is_prime):  # 현재 소수에서 한 자리만 바꾼 소수들을 반복
            if not visited[neighbor]:  # 아직 방문하지 않은 소수에 대해
                if neighbor == end:  # 이웃 소수가 목표 소수와 같으면
                    return steps + 1  # 변환 횟수를 반환
                visited[neighbor] = True  # 방문 처리
                queue.append((neighbor, steps + 1))  # 이웃 소수와 변환 횟수를 큐에 추가
    
    return "Impossible"  # 목표 소수에 도달하지 못하면 "Impossible" 반환

def main():  # 메인 함수
    T = int(input())  # 테스트 케이스의 수 입력
    is_prime = sieve_of_eratosthenes()  # 미리 1000부터 9999까지의 소수를 구함
    
    for _ in range(T):  # 각 테스트 케이스에 대해
        start, end = map(int, input().split())  # 시작 소수와 목표 소수 입력
        result = bfs(start, end, is_prime)  # BFS로 최소 변환 횟수를 계산
        print(result)  # 결과 출력

if __name__ == "__main__":  # 프로그램의 시작점
    main()  # 메인 함수 호출
